contracts/ConversionRates2.sol

```diff
@@ -0,0 +1,215 @@
+        if (!tokenData[token].enabled) return 0;
+        if (tokenControlInfo[token].minimalRecordResolution == 0) return 0; // token control info not set
+
+        // get rate update block
+        bytes32 compactData = tokenRatesCompactData[tokenData[token].compactDataArrayIndex];
+
+        uint updateRateBlock = getLast4Bytes(compactData);
+        if (currentBlockNumber >= updateRateBlock + validRateDurationInBlocks) return 0; // rate is expired
+        // check imbalance
+        int totalImbalance;
+        int blockImbalance;
+        (totalImbalance, blockImbalance) = getImbalance(token, updateRateBlock, currentBlockNumber);
+
+        // calculate actual rate
+        int imbalanceQty;
+        int extraBps;
+        int8 rateUpdate;
+        uint rate;
+
+        if (buy) {
+            // start with base rate
+            rate = tokenData[token].baseBuyRate;
+
+            // add rate update
+            rateUpdate = getRateByteFromCompactData(compactData, token, true);
+            extraBps = int(rateUpdate) * 10;
+            rate = addBps(rate, extraBps);
+
+            // compute token qty
+            qty = getTokenQty(token, qty, rate);
+            imbalanceQty = int(qty);
+            totalImbalance += imbalanceQty;
+
+            // add qty overhead
+            extraBps = executeStepFunction(tokenData[token].buyRateQtyStepFunction, int(qty));
+            rate = addBps(rate, extraBps);
+
+            // add imbalance overhead
+            extraBps = executeStepFunction(tokenData[token].buyRateImbalanceStepFunction, totalImbalance);
+            rate = addBps(rate, extraBps);
+        } else {
+            // start with base rate
+            rate = tokenData[token].baseSellRate;
+
+            // add rate update
+            rateUpdate = getRateByteFromCompactData(compactData, token, false);
+            extraBps = int(rateUpdate) * 10;
+            rate = addBps(rate, extraBps);
+
+            // compute token qty
+            imbalanceQty = -1 * int(qty);
+            totalImbalance += imbalanceQty;
+
+            // add qty overhead
+            extraBps = executeStepFunction(tokenData[token].sellRateQtyStepFunction, int(qty));
+            rate = addBps(rate, extraBps);
+
+            // add imbalance overhead
+            extraBps = executeStepFunction(tokenData[token].sellRateImbalanceStepFunction, totalImbalance);
+            rate = addBps(rate, extraBps);
+        }
+
+        if (abs(totalImbalance) >= getMaxTotalImbalance(token)) return 0;
+        if (abs(blockImbalance + imbalanceQty) >= getMaxPerBlockImbalance(token)) return 0;
+
+        return rate;
+    }
+
+    function executeStepFunction(StepFunction f, int x) internal pure returns(int) {
+        uint len = f.y.length;
+        uint ind;
+        if (x == 0) {
+            // fallback to old logics
+            for(ind = 0; ind < len; ind++) {
+                if (x <= f.x[ind]) { return f.y[ind]; }
+            }
+            return f.x[len - 1];
```

Comment:
should we add for `len` check here? or just check if `len == 0` after defined `uint len = f.y.length`